File: leetcode/Python/min-cost-to-connect-all-points.py
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        graph, pq, visited, visit = [list() for _ in range(len(points))], list(), 0, [0]*(len(points))
        sol, idx = 0, 0
        for i in range(len(points)):
            for j in range(i+1, len(points)):
                dist = int(abs(points[i][0] - points[j][0]) + abs(points[i][1] - points[j][1]))
                graph[i].append([dist, j])
                graph[j].append([dist, i])
        visited, visit[0] = visited+1, 1
        for g in graph[0]:
            heapq.heappush(pq, g)
        while pq and visited<len(points):
            dist, cur = heapq.heappop(pq)
            if visit[cur]>0:
                continue
            sol, visit[cur], visited = sol + dist , 1, visited+1
            for g in graph[cur]:
                if visit[g[1]]==0:
                    heapq.heappush(pq, g)
        return sol

File: leetcode/C++/min-cost-to-connect-all-points.cpp
class Solution {
public:
    int distance(vector<int> &a, vector<int> &b){
        int dx=(a[0]>b[0])? a[0]-b[0] : b[0]-a[0];
        int dy=(a[1]>b[1])? a[1]-b[1] : b[1]-a[1];
        return dx+dy;
    }
    int minCostConnectPoints(vector<vector<int>>& points) {
        int n = points.size(), visited=0, dist=0, cur=0, next=0, sol = 0;
        vector<vector<pair<int,int>>> graph(n);
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
        vector<int> visit(n,0);
        for(int i = 0, d=0; i < n; i++){
            for(int j = i+1; j < n; j++){
                d = distance(points[i], points[j]);
                graph[i].push_back({d, j});
                graph[j].push_back({d, i});
            }
        }
        visit[0]=visited=1;
        for(pair<int,int> &p : graph[0]) pq.push(p);
        while(!pq.empty() && visited < n){
            dist=pq.top().first; cur=pq.top().second; pq.pop();
            if(visit[cur]) continue;
            sol += dist; ++visited; visit[cur]=1;
            for(pair<int,int> &p : graph[cur]){
                if(visit[p.second]) continue;
                pq.push(p);
            }
        }
        return sol;
    }
};




/*
class Solution {
public:
    vector<int> group;
    int find(int &p){
        if(group[p]==-1) return p;
        return group[p]=find(group[p]);
    }
    int unions(int &a, int &b){
        int c=find(a);
        int d=find(b);
        if(c==d) return 0;
        group[c]=d;
        return 1;
    }
    int distance(vector<int> &a, vector<int> &b){
        int dx=(a[0]>b[0])? a[0]-b[0] : b[0]-a[0];
        int dy=(a[1]>b[1])? a[1]-b[1] : b[1]-a[1];
        return dx+dy;
    }
    int minCostConnectPoints(vector<vector<int>>& points) {
        vector<vector<int>> dp;
        int sol = 0;
        for(int i=0; i<points.size(); ++i){
            group.push_back(-1);
            for(int j=i+1; j<points.size(); ++j){
                dp.push_back({distance(points[i],points[j]),i,j});
            }
        }
        sort(dp.begin(),dp.end());
        for(int i=0,idx=0; i<dp.size()&&idx<points.size(); ++i){
            if(unions(dp[i][1],dp[i][2])) {
                ++idx; sol+=dp[i][0];
            }
        }
        return sol;
    }
};
*/